Merkle-CRDT-based datastore: Initial implementation

hsanjuan opened the pull request:
This is the initial implementation of go-ds-crdt. It satisfies the
Datastore interface and has some tests using an in-memory datastore.

hsanjuan commented:
cc @haadcode

lanzafame commented on crdt.go:
```diff
@@ -0,0 +1,532 @@
+)
+
+type Broadcaster interface {
+	Broadcast([]byte) error
```
```suggestion
	Broadcast(context.Context, []byte) error
```
Broadcast should then be responsible for serializing the context for propogation, if the implementation wishes.

hsanjuan replied:
The implementer can work around this. I'm not in favour of polluting the interfaces with contexts which have no attached meaning other than potential tracing.

lanzafame commented on crdt.go:
```diff
@@ -0,0 +1,532 @@
+
+type Broadcaster interface {
+	Broadcast([]byte) error
+	Next() ([]byte, error)
```
```suggestion
	Next() (context.Context, []byte, error)
```
Next should be responsible for the deserializing of the context from the payload, if one exists.

hsanjuan commented on .gx/lastpubver:
```diff
@@ -0,0 +1 @@
+0.0.1: QmaeLtTvTgEhHJmpNywnRDVhzqbLehdqJk5MvLDdqhWGcv
```
Not until deps support go.mod. But likely that's going to happen quite soon.

lanzafame replied:
@hsanjuan this has happened now, so could you please add it :+1:

hsanjuan replied:
done

hsanjuan commented on crdt.go:
```diff
@@ -0,0 +1,532 @@
+	Next() ([]byte, error)
+}
+
+type DAGSyncer interface {
```
I might be missing general godoc strings, yes

hsanjuan commented on crdt.go:
```diff
@@ -0,0 +1,532 @@
+		return nil
+	}
+
+	// if store.heads.Len() == 0 { // no heads? We must be syncing from scratch
```
ok, I'll think of it. We need something on this front.

hsanjuan replied:
I have a plan

hsanjuan replied:
will remove

hsanjuan commented on crdt.go:
```diff
@@ -0,0 +1,532 @@
+}
+
+type Options struct {
+	Logger logging.StandardLogger
```
nop, see https://github.com/libp2p/go-libp2p/issues/526#issuecomment-459211659 for a reasoned explanation. Functional pattern may hide complexity, but there is no complexity here to hide at the moment at least.

hsanjuan commented on crdt.go:
```diff
@@ -0,0 +1,532 @@
+	return nil
+}
+
+// walkBranch walks down a branch and applies its deltas until it arrives to a
```
`the datastore's deltas` is confusing.  It's more the node's deltas.

hsanjuan commented on ipld.go:
```diff
@@ -0,0 +1,165 @@
+	// launch maximum ng.workers. Do not launch more workers than cids
+	// though.
+	for w := 0; w < ng.workers && w < len(cids); w++ {
+		go func() {
```
The DAGSyncer should handle that? If a context is cancelled it should decide what to do with the Get operation it's performing

lanzafame replied:
But the `DAGSyncer` has no way of communicating to the `w` number of workers that `GetMany` has spun up?

hsanjuan replied:
Removed this by requiring passing in a full DAGSyncer, so it doesn't apply

hsanjuan commented on ipld.go:
```diff
@@ -0,0 +1,165 @@
+// 	return ng.rpcClient.CallContext(
+// 		ctx,
+// 		"",
+// 		"Cluster",
```
yes, everything.. this needs removal

hsanjuan commented on set.go:
```diff
@@ -0,0 +1,416 @@
+	// have a value in the store.
+
+	// An element having a value in the store implies that it is in
+	// "elems".
```
Not sure this is correct. Value is never removed from elems. It is removed from the store when all values in elems also happen to be tombstoned.

lanzafame replied:
I am not sure it is correct either, but the existing comment doesn't make it clear. I just suggested what I had derived from reading the code.

hsanjuan replied:
rewrote

hsanjuan commented on set.go:
```diff
@@ -0,0 +1,416 @@
+	return retChan
+}
+
+// InSet returns if we have a Cid. For it we must have at least one element
```
Somethings funny with grammar in the suggestion

lanzafame replied:
Corrected suggestion.

hsanjuan commented on set.go:
```diff
@@ -0,0 +1,416 @@
+	return s.store.Has(k)
+}
+
+// currently unused
```
I'll leave it a bit longer as it may become useful

lanzafame commented on set.go:
```diff
@@ -0,0 +1,416 @@
+// sets a value if priority is higher. When equal, it sets if the
+// value is lexicographically higher than the current value.
+func (s *set) setValue(key string, value []byte, prio uint64) error {
+	curPrio, err := s.getPrio(key)
```
```suggestion
	curPrio, err := s.getPriority(key)
```

lanzafame commented on set.go:
```diff
@@ -0,0 +1,416 @@
+	}
+
+	// store priority
+	return s.setPrio(key, prio)
```
```suggestion
	return s.setPriority(key, prio)
```

hsanjuan commented:
@lanzafame I have addressed all comments (except the two in disagreement). I have added some things like DAGSyncer being mostly a full DAGService. My idea now is to fully skip IPFS to store crdt blocks and store them directly cluster, and do bitswap etc among cluster peers. This will make things way better (reduced dht size, working bitswap sessions) and the effort is small.

I think we could merge this soon and I'll do the re-broadcasting on another PR (also need to add some info logging).

(tests are funky but I think it's because of not waiting long enough for things to propagage)

kishansagathiya commented on set.go:
```diff
@@ -0,0 +1,409 @@
+// Element retrieves the value of an element from the CRDT set.
+func (s *set) Element(key string) ([]byte, error) {
+	// We can only GET an element if it's part of the Set (in
+	// "elems" and not in "tombstones").
```
writing like this signals that `elems` and `tombstones` are variable names
```suggestion
	// elements and not in tombstones).
```

hsanjuan replied:
fixed this leaving quotes

kishansagathiya commented on set.go:
```diff
@@ -0,0 +1,409 @@
+	// "elems" and not in "tombstones").
+
+	// As an optimization:
+	// * If the element has a value in the store it means:
```
```suggestion
	// * If the key has a value in the store it means:
```

hsanjuan replied:
ok

kishansagathiya commented on set.go:
```diff
@@ -0,0 +1,409 @@
+	// * If the element has a value in the store it means:
+	//   -> It occurs at least once in "elems"
+	//   -> It may or not be tombstoned
+	// * If the element does not have a value in the store:
```
```suggestion
	// * If the key does not have a value in the store:
```
the comment can be modified accordingly

hsanjuan replied:
ok

kishansagathiya commented on set.go:
```diff
@@ -0,0 +1,409 @@
+	//   -> In both cases the element "does not exist".
+
+	valueK := s.valueKey(key)
+	value, err := s.store.Get(valueK)
```
`s.store.Get(valueK)`-->`store.set.Element(key.String())`

both `store` are same here, seems like an infinite loop

https://github.com/ipfs/go-ds-crdt/blob/cdaa501f3c4295e7d86c85d8c1349956f69f2890/crdt.go#L274

hsanjuan replied:
No, `s.store` in the Set type is the undelying datastore, and `store` in `crdt.go` is the cdrt datastore that we are implementing.

hsanjuan commented on set.go:
```diff
@@ -0,0 +1,409 @@
+	namespace ds.Key
+}
+
+func newCRDTSet(d ds.Datastore, namespace ds.Key) *set {
```
I don't see much advantage

hsanjuan commented on set.go:
```diff
@@ -0,0 +1,409 @@
+				Value: value,
+			},
+		},
+		Tombstones: nil,
```
Seems the comment is old. Updated it.

> This could not be a method of set and would still work, so why?

Because it is way clearer that you are getting the delta to add to a set, just like `Rmv`, regardless of whether the set is needed internally or not. 

> Looking at the code it doesn't seem like it is not adding, but rather it returns a new pb.Delta

This is how the Add operation is defined in delta-crdts. Deltas need to be merged later.

hsanjuan commented on set.go:
```diff
@@ -0,0 +1,409 @@
+
+	for r := range results.Next() {
+		if r.Error != nil {
+			return delta, err
```
I don't see a problem here.

hsanjuan replied:
delta is partially built at this point. In the previous errors the delta does not hold any items. I don't have a strong preference here, thus I'd like to leave it as is.

hsanjuan commented on set.go:
```diff
@@ -0,0 +1,409 @@
+// /namespace/<key>/priority
+func (s *set) priorityKey(key string) ds.Key {
+	return s.keyPrefix(key).ChildString(prioritySuffix)
+}
```
yeah comments intend to be meaningful. Actual keys are shortened.

hsanjuan commented on package.json:
```diff
@@ -0,0 +1,30 @@
+  },
+  "gxDependencies": [
+    {
+      "author": "why",
```
This is how that dependency was gxed. It doesn't matter

hsanjuan commented on pb/delta.proto:
```diff
@@ -0,0 +1,15 @@
+
+message Delta {
+  repeated Element elements = 1;
+  repeated Element tombstones = 2;
```
A delta is just a Set like the full set, but usually a part of it. It doesn't necessarily need to carry new elements, but will usually do so.

kishansagathiya commented on crdt.go:
```diff
@@ -0,0 +1,531 @@
+		Elements:   elems,
+		Tombstones: store.curDelta.GetTombstones(),
+		Priority:   store.curDelta.GetPriority(),
+	}
```
isn't this same as
```go
store.curDelta.Elements = elems
```

hsanjuan replied:
store.curDelta might be `nil`, so, not exactly the same. This approach is safe on that regard.

kishansagathiya commented on crdt.go:
```diff
@@ -0,0 +1,531 @@
+}
+
+// IsThreadSafe declares that this datastore implementation is thread-safe.
+func (store *Datastore) IsThreadSafe() {}
```
Fill this or add a `TODO` ?

hsanjuan replied:
There is nothing todo. That's just how a datastore implements the ThreadSafe interface for datastores.

kishansagathiya commented on crdt_test.go:
```diff
@@ -0,0 +1,443 @@
+func (tl *testLogger) Warningf(format string, args ...interface{}) {
+	args = append([]interface{}{tl.name}, args...)
+	tl.l.Warningf("%s "+format, args...)
+}
```
testLogger seems like something that can move to a separate file, say `test_util.go`

hsanjuan replied:
I would not move it to a non test file. Moving it to a separate test file makes me connsider moving also mockBroadcaster or mockDAGSync. In the end I don't think there is any significant advantage in having it live in a separate place (and it makes you need to open more files).

hsanjuan commented:
Ok, let's land this and iterate.
